codesign/lib/src/utils.dart

```diff
@@ -2,7 +2,71 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:io';
+
 import 'package:args/args.dart';
+import 'package:codesign/src/file_codesign_visitor.dart';
+import 'package:process/process.dart';
+
+/// helper function to generate unique next IDs.
+int id = 0;
+int get nextId {
+  final int currentKey = id;
+  id += 1;
+  return currentKey;
+}
+
+enum FileType { folder, zip, binary, other }
+
+Future<void> unzip(FileSystemEntity inputZip, Directory outDir, ProcessManager processManager) async {
+  await processManager.run(
+    <String>[
+      'unzip',
+      inputZip.absolute.path,
+      '-d',
+      outDir.absolute.path,
+    ],
+  );
+  logger!.info('The downloaded file is unzipped from ${inputZip.absolute.path} to ${outDir.absolute.path}\n');
+}
+
+Future<void> zip(Directory inDir, FileSystemEntity outputZip, ProcessManager processManager) async {
+  await processManager.run(
+    <String>[
+      'zip',
+      '--symlinks',
+      '--recurse-paths',
+      outputZip.absolute.path,
+      // use '.' so that the full absolute path is not encoded into the zip file
+      '.',
+      '--include',
+      '*',
+    ],
+    workingDirectory: inDir.absolute.path,
+  );
+}
+
+/// Check mime-type of file at [filePath] to determine if it is a directory.
+FileType getFileType(String filePath, ProcessManager processManager) {
+  final ProcessResult result = processManager.runSync(
+    <String>[
+      'file',
+      '--mime-type',
+      '-b', // is binary
+      filePath,
+    ],
+  );
+  String output = result.stdout as String;
+  if (output.contains('inode/directory')) {
+    return FileType.folder;
+  } else if (output.contains('application/zip')) {
+    return FileType.zip;
+  } else if (output.contains('application/x-mach-binary')) {
+    return FileType.binary;
+  } else {
+    return FileType.other;
```

Comment:
In other cases, should we be throwing an exception instead? Silently failing will make it harder to debug in the future

Reply:
umm the artifact files have many different types, `application/octet-stream` for frontend_server.dart.snapshot, `text/x-objective-c` for FlutterAppDelegate.h, `text/plain` for module.modulemap, `text/xml` for info.plist, and many others. We would have to exclude all these types if we want to throw an exception. I might be wrong, code sign isn't interested in the majority of file types but maybe we don't want to crash on them.

Reply:
This would make a good comment in the code

Reply:
Sounds good, added comment